File: android/mesh/src/main/java/io/godstone/mesh/delivery/SqliteDeliveryJournal.kt
package io.godstone.mesh.delivery

import io.godstone.mesh.store.StoreDb

// Stage 4C.1 / C6.1 -- production [DeliveryJournal] backed by the SAME SQLite DB
// as the held-frames store. The delivery lifecycle state, the ACK mode, and the
// intended recipient live in ONE row keyed by msg_id, so the ACK authenticity
// decision (C2) binds to the recipient recorded in durable outbound state
// INDEPENDENT of the ACK frame (ADR-005: do not claim an ACK is from the
// intended recipient unless the expected recipient comes from durable outbound
// state).
//
// Each mutation is a single atomic SQL statement (see StoreSchema): [insert]
// creates the row (INSERT ... ON CONFLICT DO NOTHING), [updateState] advances
// only the state column (UPDATE SET state WHERE msg_id, preserving ack_mode +
// expected_recipient), [clear] drops the row. The expected recipient is IMMUTABLE
// post-creation -- there is no recipient-only write. A crash between
// DeliveryTracker's read and write leaves the LAST persisted state on disk (the
// write did not commit), which is the crash-safe semantics ADR-005 requires.
// (CAS-hardened transitions arrive in C6.4.)
//
// The int state / ack_mode codes are the cross-platform persistence contract
// (NOT the Kotlin enum ordinals), so Android and iOS agree even if their enum
// orders ever diverge. An unknown state or ack_mode code, or a row that
// violates the C6.1 binding invariant, decodes to [DeliveryLookup.Corrupt] --
// fail closed (C6.5), never silently to UNAVAILABLE. Mirrors
// `SqliteDeliveryJournal` on iOS (byte-identical schema + SQL + codes).
internal class SqliteDeliveryJournal(private val db: StoreDb) : DeliveryJournal {

    override fun read(msgId: ByteArray): DeliveryLookup {
        val row = db.readDelivery(msgId) ?: return DeliveryLookup.NotFound
        val state = DeliveryState.fromCode(row.state) ?: return DeliveryLookup.Corrupt
        val ackMode = AckMode.fromCode(row.ackMode) ?: return DeliveryLookup.Corrupt
        // C6.1 binding invariant (mirrors the schema CHECK): NONE -> null
        // recipient; SINGLE_RECIPIENT -> 16-byte recipient. A row that violates
        // it (e.g. manually mutated) is corrupt -- fail closed.
        if (!bindingConsistent(ackMode, row.expectedRecipient)) return DeliveryLookup.Corrupt
        return DeliveryLookup.Found(
            DeliveryRecord(msgId, state, ackMode, row.expectedRecipient),
        )
    }

    override fun insert(msgId: ByteArray, ackMode: AckMode, expectedRecipient: ByteArray?): Boolean =
        db.insertDelivery(msgId, DeliveryState.QUEUED_DURABLY.code, ackMode.code, expectedRecipient)

    override fun updateState(msgId: ByteArray, state: DeliveryState): Int =
        db.updateDeliveryState(msgId, state.code)

    override fun clear(msgId: ByteArray) {
        db.clearDelivery(msgId)
    }

    private fun bindingConsistent(ackMode: AckMode, expectedRecipient: ByteArray?): Boolean =
        when (ackMode) {
            AckMode.NONE -> expectedRecipient == null
            AckMode.SINGLE_RECIPIENT -> expectedRecipient != null && expectedRecipient.size == 16
        }
}

File: ios/Godstone/Sources/GodstoneMesh/SqliteDeliveryJournal.swift
import Foundation

// Stage 4C.1 / C6.1 -- production `DeliveryJournal` backed by the SAME SQLite DB
// as the held-frames store. The delivery lifecycle state, the ACK mode, and the
// intended recipient live in ONE row keyed by msg_id, so the ACK authenticity
// decision (C2) binds to the recipient recorded in durable outbound state
// INDEPENDENT of the ACK frame (ADR-005: do not claim an ACK is from the
// intended recipient unless the expected recipient comes from durable outbound
// state).
//
// Each mutation is a single atomic SQL statement (see StoreSchema): `insert`
// creates the row (INSERT ... ON CONFLICT DO NOTHING), `updateState` advances
// only the state column (UPDATE SET state WHERE msg_id, preserving ack_mode +
// expected_recipient), `clear` drops the row. The expected recipient is IMMUTABLE
// post-creation -- there is no recipient-only write. A crash between
// DeliveryTracker's read and write leaves the LAST persisted state on disk (the
// write did not commit), which is the crash-safe semantics ADR-005 requires.
// (CAS-hardened transitions arrive in C6.4.)
//
// The int state / ack_mode codes are the cross-platform persistence contract
// (DeliveryState.code / AckMode.rawValue, NOT the Swift enum order), so Android
// and iOS agree even if their enum orders ever diverge. An unknown state or
// ack_mode code, or a row that violates the C6.1 binding invariant, decodes to
// `DeliveryLookup.corrupt` -- fail closed (C6.5), never silently to unavailable.
// Mirrors `SqliteDeliveryJournal` on Android (byte-identical schema + SQL +
// codes).

/// Production `DeliveryJournal` over the same SQLite DB as `SqliteMessageStore`'s
/// held-frames table. One row holds the delivery state, ack mode, and intended
/// recipient, so the row read at acknowledge time is the row written at enqueue
/// time. The separate `ExpectedRecipientStore` seam was removed in C6.1: the
/// journal IS the durable record.
public final class SqliteDeliveryJournal: DeliveryJournal {
    private let store: SqliteMessageStore

    public init(_ store: SqliteMessageStore) {
        self.store = store
    }

    public func read(_ msgId: Data) -> DeliveryLookup {
        guard let row = store.readDelivery(msgId) else { return .notFound }
        guard let state = DeliveryState.fromCode(row.state) else { return .corrupt }
        guard let ackMode = AckMode.fromCode(row.ackMode) else { return .corrupt }
        // C6.1 binding invariant (mirrors the schema CHECK): none -> nil
        // recipient; singleRecipient -> 16-byte recipient. A row that violates it
        // (e.g. manually mutated) is corrupt -- fail closed.
        guard bindingConsistent(ackMode: ackMode, expectedRecipient: row.expectedRecipient) else {
            return .corrupt
        }
        return .found(DeliveryRecord(msgId: msgId, state: state, ackMode: ackMode,
                                     expectedRecipientNodeId: row.expectedRecipient))
    }

    public func insert(_ msgId: Data, ackMode: AckMode, expectedRecipient: Data?) -> Bool {
        store.insertDelivery(msgId,
                             stateOrdinal: DeliveryState.queuedDurably.code,
                             ackModeOrdinal: ackMode.rawValue,
                             expectedRecipient: expectedRecipient)
    }

    public func updateState(_ msgId: Data, _ state: DeliveryState) -> Int {
        store.updateDeliveryState(msgId, stateOrdinal: state.code)
    }

    public func clear(_ msgId: Data) {
        store.clearDelivery(msgId)
    }

    /// C6.1 invariant: none -> nil recipient; singleRecipient -> 16-byte recipient.
    private func bindingConsistent(ackMode: AckMode, expectedRecipient: Data?) -> Bool {
        switch ackMode {
        case .none: return expectedRecipient == nil
        case .singleRecipient:
            guard let r = expectedRecipient else { return false }
            return r.count == 16
        }
    }
}